tools/swift-inspect/Sources/SwiftInspectLinux/PTrace.swift

```diff
@@ -0,0 +1,151 @@
+  }
+
+  let pid: pid_t
+
+  // Initializing a PTrace instance attaches to the target process, waits for
+  // it to stop, and leaves it in a stopped state. The caller may resume the
+  // process by calling cont().
+  public init(process pid: pid_t) throws {
+    guard ptrace_attach(pid) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_ATTACH, pid: pid)
+    }
+
+    while true {
+      var status: Int32 = 0
+      let result = waitpid(pid, &status, 0)
+      guard result != -1 else {
+        if get_errno() == EINTR { continue }
+        throw PTraceError.waitFailure(pid: pid)
+      }
+
+      precondition(pid == result, "waitpid returned unexpected value \(result)")
+
+      if wIfStopped(status) { break }
+    }
+
+    self.pid = pid
+  }
+
+  deinit { _ = ptrace_detach(self.pid) }
+
+  public func cont() throws {
+    guard ptrace_cont(self.pid) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_CONT, pid: self.pid)
+    }
+  }
+
+  public func interrupt() throws {
+    guard ptrace_interrupt(self.pid) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_INTERRUPT, pid: self.pid)
+    }
+  }
+
+  public func getSigInfo() throws -> siginfo_t {
+    var sigInfo = siginfo_t()
+    guard ptrace_getsiginfo(self.pid, &sigInfo) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_GETSIGINFO, pid: self.pid)
+    }
+
+    return sigInfo
+  }
+
+  public func pokeData(addr: UInt64, value: UInt64) throws {
+    guard ptrace_pokedata(self.pid, UInt(addr), UInt(value)) != -1 else {
+      throw PTraceError.ptraceFailure(PTRACE_POKEDATA, pid: self.pid)
+    }
+  }
+
+  public func getRegSet() throws -> RegisterSet {
+    var regSet = RegisterSet()
+    try withUnsafeMutableBytes(of: &regSet) {
+      var vec = iovec(iov_base: $0.baseAddress!, iov_len: MemoryLayout<RegisterSet>.size)
+      guard ptrace_getregset(self.pid, NT_PRSTATUS, &vec) != -1 else {
+        throw PTraceError.ptraceFailure(PTRACE_GETREGSET, pid: self.pid)
+      }
+    }
+
+    return regSet
+  }
+
+  public func setRegSet(regSet: RegisterSet) throws {
+    var regSetCopy = regSet
+    try withUnsafeMutableBytes(of: &regSetCopy) {
+      var vec = iovec(iov_base: $0.baseAddress!, iov_len: MemoryLayout<RegisterSet>.size)
+      guard ptrace_setregset(self.pid, NT_PRSTATUS, &vec) != -1 else {
+        throw PTraceError.ptraceFailure(PTRACE_SETREGSET, pid: self.pid)
+      }
+    }
+  }
+
+  // Call the function at the specified address in the attached process. Caller
+  // may pass up to six 8-byte arguments. The optional callback is invoked when
+  // the process is stopped with a SIGTRAP signal. In this case, the caller is
+  // responsible for taking action on the signal.
+  public func callRemoteFunction(
+    at address: UInt64, with args: [UInt64] = [], onTrap callback: (() throws -> Void)? = nil
+  ) throws -> UInt64 {
+    precondition(args.count <= 6, "callRemoteFunction supports max of 6 arguments")
```

Comment:
Why 6? X64 does 6 (rsi, rdi, rcx, rdx r8, r9) but AAPCS suggests 8 (r0, r1, r2, r3, r4, r5, r6, r7). x86 actually does 4 but only in certain cases. So, perhaps it should be 0?

Reply:
Because it was the smaller of 6 and 8. The function with the most arguments we call is `mmap` which takes 6 args so it doesn't actually matter in this context.

Regardless, I will move this assertion into the arch-specific setup functions and enforce 6 for x86_64 and 8 for arm64 instead.